### ISSUE TITLE
Occasional test suite failures in t/e2e/aeval_merge.t and t/e2e/eval_merge.t

### ISSUE THREAD

**Issue by xtaran** [pjcj/Devel--Cover#143]
Hi Paul,

as already reported by private mail when discovered, here's a more complete and official bug report.

I first discovered this on the Debian build daemons [on i386](https://buildd.debian.org/status/fetch.php?pkg=libdevel-cover-perl&arch=i386&ver=1.21-1&stamp=1443050049) and [hurd-i386](https://buildd.debian.org/status/fetch.php?pkg=libdevel-cover-perl&arch=hurd-i386&ver=1.21-1&stamp=1443051175).

I could not reproduce it immediately on a i386, but later after running `while make test; do sleep 1; done` for some time, I was able to reproduce the issue after ca. 3 minutes and 8 successful runs. (Which is more or less the same percentage of failures as on the Debian build daemons.)

The failing part, example taken [from `t/e2e/aeval_merge.t` on i386](https://buildd.debian.org/status/fetch.php?pkg=libdevel-cover-perl&arch=i386&ver=1.21-1&stamp=1443050049):

```
| 117|  '10                                      print "E3\\n";                                                                 |  '10                                      print "E3\\n";                                                                 |
| 118|',                                                                                                                        |',                                                                                                                        |
| 119|  '11                                                                                                                     |  '11                                                                                                                     |
| 120|',                                                                                                                        |',                                                                                                                        |
* 121|  '12             2                    2   sub E3 { print "E3::E3\\n" }                                                   |  '12             1                    1   sub E3 { print "E3::E3\\n" }                                                   *
| 122|',                                                                                                                        |',                                                                                                                        |
| 123|  '13                                                                                                                     |  '13                                                                                                                     |
| 124|',                                                                                                                        |',                                                                                                                        |
| 125|  '14                                      1                                                                              |  '14                                      1                                                                              |
| 126|',                                                                                                                        |',                                                                                                                        |
| 127|  '                                                                                                                       |  '                                                                                                                       |
| 128|',                                                                                                                        |',                                                                                                                        |
| 129|  '                                                                                                                       |  '                                                                                                                       |
| 130|',                                                                                                                        |',                                                                                                                        |
| 131|  'Covered Subroutines                                                                                                    |  'Covered Subroutines                                                                                                    |
| 132|',                                                                                                                        |',                                                                                                                        |
| 133|  '-                                                                                                                      |  '-                                                                                                                      |
| 134|',                                                                                                                        |',                                                                                                                        |
| 135|  '                                                                                                                       |  '                                                                                                                       |
| 136|',                                                                                                                        |',                                                                                                                        |
| 137|  'Subroutine Count Location                                                                                              |  'Subroutine Count Location                                                                                              |
| 138|',                                                                                                                        |',                                                                                                                        |
| 139|  '- ----- --------------                                                                                                 |  '- ----- --------------                                                                                                 |
| 140|',                                                                                                                        |',                                                                                                                        |
* 141|  'E3             2 tests/E3.pm:12                                                                                        |  'E3             1 tests/E3.pm:12                                                                                        *
```

In Debian we track this issue as [#800424](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=800424)

**Comment by pjcj** [comment]
Thanks very much for reporting this.

I have seen similar problems that have come down to which other modules are installed.  But I'm not sure whether that is relevant here.

Will need to look into it further.

**Comment by Grueslayer** [comment]
This problem is reproduceable ALL times running Strawberry Perl v5.22.0 MSWin32-x86-multi-thread-64int but for E4

```
| 120|',                                                                                                                        |',                                                                                                                        |
* 121|  '12             2                    2   sub E4 { print "E4::E4\\n" }                                                   |  '12             1                    1   sub E4 { print "E4::E4\\n" }                                                   *
| 122|',                                                                                                                        |',                                                                                                                        |

...

| 140|',                                                                                                                        |',                                                                                                                        |
* 141|  'E4             2 tests/E4.pm:12                                                                                        |  'E4             1 tests/E4.pm:12                                                                                        *
| 142|',                                                                                                                        |',                                                                                                                        |

```

**Comment by ntyni** [comment]
I can get this to happen quite reliably on Debian sid/i386 (=32-bit). Digging into it, it looks like Perl allocates the same memory address for the print statement ops in E2::E2() and E3::E3(), resulting in a collision in get_key() and the former getting counted twice.

I assume Perl is just better than expected at reusing memory here, although I'm not really sure if it's breaking some guarantee that I'm not aware of.

Adding file and line information into the key seems like a good way to prevent such collisions. Please see https://github.com/ntyni/Devel--Cover/commit/2fccbe2e2566377eb4be1e7318f0db6d9387bc9f   which fixes the issue for me without breaking any existing tests.

Let me know if you'd like debugging traces or something like that.

**Comment by ntyni** [comment]
Ping? We're also seeing this on the 'armhf' Debian architecture, so it's not even i386 specific.

**Comment by xtaran** [comment]
Hi Paul,

IIRC I've read somewhere (maybe in one of the other issues) that you're attending this year's Perl QA Hackathon. Any chance that you find time some time there to look at @ntyni's patch for this issue?

While amd64 (x86_64) didn't show this issue with 1.22, it showed up again in Debian on [i386 (x86 32 bits, second build only)](https://tests.reproducible-builds.org/rb-pkg/unstable/i386/libdevel-cover-perl.html) and [on armhf](https://tests.reproducible-builds.org/rb-pkg/unstable/armhf/libdevel-cover-perl.html).

TIA!

**Comment by pjcj** [comment]
It's true - I am at QAH! I will set aside some time to look at this - thanks for the reminder.

**Comment by mapreri** [comment]
@pjcj another friendly ping here :)

**Comment by pjcj** [comment]
I know. I'm sorry :( I have so little time for Devel::Cover nowadays.

But, I'm so late that there is another QAH (albeit with a different name) coming up soon...

**Comment by pjcj** [comment]
I have eventually got around to looking at this - sorry for the delay.

I have not been able to reproduce the problem myself, but the analysis and solution sound feasible and look good.  I have taken the fix at ntyni@2fccbe2, with minor changes, and applied it to the master branch with c0e8160.  The results look good to me, but I don't know whether the problem is now fixed.

Would anyone be able to take the master branch and see whether the problem is now fixed?

Thanks to everyone who has helped with this ticket so far.

**Comment by gregoa** [comment]
I've uploaded 1.25 to experimental right now, build results will show up at
https://buildd.debian.org/status/package.php?p=libdevel-cover-perl&suite=experimental
soon.

**Comment by pjcj** [comment]
How are things looking here?  My experience suggests that problem may be solved.

**Comment by ntyni** [comment]
On Thu, Sep 14, 2017 at 07:02:46PM +0000, Paul Johnson wrote:
> How are things looking here?  My experience suggests that problem may be solved.

Hi, all the build logs I can find on the Debian side also suggest that
this is fixed with 1.25, as expected.

Many thanks!
-- 
Niko

**Comment by pjcj** [comment]
Excellent - thanks to everyone involved.
